File: src/NodeParser/InterfaceAndClassNodeParser.ts
import ts, { PropertyName } from "typescript";
import { Context, NodeParser } from "../NodeParser";
import { SubNodeParser } from "../SubNodeParser";
import { ArrayType } from "../Type/ArrayType";
import { BaseType } from "../Type/BaseType";
import { NeverType } from "../Type/NeverType";
import { ObjectProperty, ObjectType } from "../Type/ObjectType";
import { ReferenceType } from "../Type/ReferenceType";
import { isNodeHidden } from "../Utils/isHidden";
import { isPublic, isStatic } from "../Utils/modifiers";
import { getKey } from "../Utils/nodeKey";

export class InterfaceAndClassNodeParser implements SubNodeParser {
    public constructor(
        protected typeChecker: ts.TypeChecker,
        protected childNodeParser: NodeParser,
        protected readonly additionalProperties: boolean
    ) {}

    public supportsNode(node: ts.InterfaceDeclaration | ts.ClassDeclaration): boolean {
        return node.kind === ts.SyntaxKind.InterfaceDeclaration || node.kind === ts.SyntaxKind.ClassDeclaration;
    }

    public createType(
        node: ts.InterfaceDeclaration | ts.ClassDeclaration,
        context: Context,
        reference?: ReferenceType
    ): BaseType {
        if (node.typeParameters?.length) {
            node.typeParameters.forEach((typeParam) => {
                const nameSymbol = this.typeChecker.getSymbolAtLocation(typeParam.name)!;
                context.pushParameter(nameSymbol.name);

                if (typeParam.default) {
                    const type = this.childNodeParser.createType(typeParam.default, context);
                    context.setDefault(nameSymbol.name, type);
                }
            });
        }

        const id = this.getTypeId(node, context);
        if (reference) {
            reference.setId(id);
            reference.setName(id);
        }

        const properties = this.getProperties(node, context);

        if (properties === undefined) {
            return new NeverType();
        }

        const additionalProperties = this.getAdditionalProperties(node, context);

        // When type only extends Array or ReadonlyArray then create an array type instead of an object type
        if (properties.length === 0 && additionalProperties === false) {
            const arrayItemType = this.getArrayItemType(node);
            if (arrayItemType) {
                return new ArrayType(this.childNodeParser.createType(arrayItemType, context));
            }
        }

        return new ObjectType(id, this.getBaseTypes(node, context), properties, additionalProperties);
    }

    /**
     * If specified node extends Array or ReadonlyArray and nothing else then this method returns the
     * array item type. In all other cases null is returned to indicate that the node is not a simple array.
     *
     * @param node - The interface or class to check.
     * @return The array item type if node is an array, null otherwise.
     */
    protected getArrayItemType(node: ts.InterfaceDeclaration | ts.ClassDeclaration): ts.TypeNode | null {
        if (node.heritageClauses && node.heritageClauses.length === 1) {
            const clause = node.heritageClauses[0];
            if (clause.types.length === 1) {
                const type = clause.types[0];
                const symbol = this.typeChecker.getSymbolAtLocation(type.expression);
                if (symbol && (symbol.name === "Array" || symbol.name === "ReadonlyArray")) {
                    const typeArguments = type.typeArguments;
                    if (typeArguments?.length === 1) {
                        return typeArguments[0];
                    }
                }
            }
        }
        return null;
    }

    protected getBaseTypes(node: ts.InterfaceDeclaration | ts.ClassDeclaration, context: Context): BaseType[] {
        if (!node.heritageClauses) {
            return [];
        }

        return node.heritageClauses.reduce(
            (result: BaseType[], baseType) => [
                ...result,
                ...baseType.types.map((expression) => this.childNodeParser.createType(expression, context)),
            ],
            []
        );
    }

    protected getProperties(
        node: ts.InterfaceDeclaration | ts.ClassDeclaration,
        context: Context
    ): ObjectProperty[] | undefined {
        let hasRequiredNever = false;

        const properties = (node.members as ts.NodeArray<ts.TypeElement | ts.ClassElement>)
            .reduce(
                (members, member) => {
                    if (ts.isConstructorDeclaration(member)) {
                        const params = member.parameters.filter((param) =>
                            ts.isParameterPropertyDeclaration(param, param.parent)
                        ) as ts.ParameterPropertyDeclaration[];
                        members.push(...params);
                    } else if (ts.isPropertySignature(member) || ts.isPropertyDeclaration(member)) {
                        members.push(member);
                    }
                    return members;
                },
                [] as (ts.PropertyDeclaration | ts.PropertySignature | ts.ParameterPropertyDeclaration)[]
            )
            .filter((member) => isPublic(member) && !isStatic(member) && !isNodeHidden(member))
            .reduce((entries, member) => {
                let memberType: ts.Node | undefined = member.type;

                // Use the type checker if the member has no explicit type
                // Ignore members without an initializer. They have no useful type.
                if (memberType === undefined && (member as ts.PropertyDeclaration)?.initializer !== undefined) {
                    const type = this.typeChecker.getTypeAtLocation(member);
                    memberType = this.typeChecker.typeToTypeNode(type, node, ts.NodeBuilderFlags.NoTruncation);
                }

                if (memberType !== undefined) {
                    return [...entries, { member, memberType }];
                }
                return entries;
            }, [])
            .map(
                ({ member, memberType }) =>
                    new ObjectProperty(
                        this.getPropertyName(member.name),
                        this.childNodeParser.createType(memberType, context),
                        !member.questionToken
                    )
            )
            .filter((prop) => {
                const type = prop.getType();
                if (prop.isRequired() && type instanceof NeverType) {
                    hasRequiredNever = true;
                }
                return !(type instanceof NeverType);
            });

        if (hasRequiredNever) {
            return undefined;
        }

        return properties;
    }

    protected getAdditionalProperties(
        node: ts.InterfaceDeclaration | ts.ClassDeclaration,
        context: Context
    ): BaseType | boolean {
        const indexSignature = (node.members as ts.NodeArray<ts.NamedDeclaration>).find(ts.isIndexSignatureDeclaration);
        if (!indexSignature) {
            return this.additionalProperties;
        }

        return this.childNodeParser.createType(indexSignature.type, context) ?? this.additionalProperties;
    }

    protected getTypeId(node: ts.Node, context: Context): string {
        const nodeType = ts.isInterfaceDeclaration(node) ? "interface" : "class";
        return `${nodeType}-${getKey(node, context)}`;
    }

    protected getPropertyName(propertyName: PropertyName): string {
        if (propertyName.kind === ts.SyntaxKind.ComputedPropertyName) {
            const symbol = this.typeChecker.getSymbolAtLocation(propertyName);
            if (symbol) {
                return symbol.getName();
            }
        }
        return propertyName.getText();
    }
}

File: internal/parser/interface_and_class_node_parser.go
package parser

// Port of src/NodeParser/InterfaceAndClassNodeParser.ts.

import (
	"github.com/microsoft/typescript-go/shim/ast"
	"github.com/microsoft/typescript-go/shim/checker"

	"github.com/vega/ts-json-schema-generator/internal/tsutils"
	"github.com/vega/ts-json-schema-generator/internal/types"
)

type InterfaceAndClassNodeParser struct {
	typeChecker          *checker.Checker
	childNodeParser      NodeParser
	additionalProperties bool
}

func NewInterfaceAndClassNodeParser(typeChecker *checker.Checker, childNodeParser NodeParser, additionalProperties bool) *InterfaceAndClassNodeParser {
	return &InterfaceAndClassNodeParser{
		typeChecker:          typeChecker,
		childNodeParser:      childNodeParser,
		additionalProperties: additionalProperties,
	}
}

func (p *InterfaceAndClassNodeParser) SupportsNode(node *ast.Node) bool {
	return node.Kind == ast.KindInterfaceDeclaration || node.Kind == ast.KindClassDeclaration
}

func (p *InterfaceAndClassNodeParser) CreateType(node *ast.Node, context *Context, reference *types.ReferenceType) types.Type {
	// Note: this mutates the caller's context, zipping parameter names onto
	// the type arguments already pushed by TypeReferenceNodeParser.
	for _, typeParam := range node.TypeParameters() {
		nameSymbol := tsutils.GetSymbolAtLocation(p.typeChecker, typeParam.Name())
		context.PushParameter(nameSymbol.Name)

		if defaultType := typeParam.AsTypeParameterDeclaration().DefaultType; defaultType != nil {
			t := p.childNodeParser.CreateType(defaultType, context, nil)
			context.SetDefault(nameSymbol.Name, t)
		}
	}

	id := p.getTypeId(node, context)
	if reference != nil {
		reference.SetID(id)
		reference.SetName(id)
	}

	properties, hasRequiredNever := p.getProperties(node, context)
	if hasRequiredNever {
		return &types.NeverType{}
	}

	additionalProperties := p.getAdditionalProperties(node, context)

	// When the type only extends Array or ReadonlyArray then create an array
	// type instead of an object type.
	if len(properties) == 0 {
		if apBool, ok := additionalProperties.(bool); ok && !apBool {
			if arrayItemType := p.getArrayItemType(node); arrayItemType != nil {
				return &types.ArrayType{Item: p.childNodeParser.CreateType(arrayItemType, context, nil)}
			}
		}
	}

	return types.NewObjectType(id, p.getBaseTypes(node, context), properties, additionalProperties, false)
}

// getArrayItemType returns the array item type if the node extends Array or
// ReadonlyArray and nothing else, nil otherwise.
func (p *InterfaceAndClassNodeParser) getArrayItemType(node *ast.Node) *ast.Node {
	clauses := heritageClausesOf(node)
	if len(clauses) != 1 {
		return nil
	}
	clauseTypes := clauses[0].AsHeritageClause().Types.Nodes
	if len(clauseTypes) != 1 {
		return nil
	}
	expr := clauseTypes[0]
	symbol := tsutils.GetSymbolAtLocation(p.typeChecker, expr.AsExpressionWithTypeArguments().Expression)
	if symbol == nil || (symbol.Name != "Array" && symbol.Name != "ReadonlyArray") {
		return nil
	}
	typeArguments := expr.TypeArguments()
	if len(typeArguments) != 1 {
		return nil
	}
	return typeArguments[0]
}

func (p *InterfaceAndClassNodeParser) getBaseTypes(node *ast.Node, context *Context) []types.Type {
	var result []types.Type
	for _, clause := range heritageClausesOf(node) {
		for _, expression := range clause.AsHeritageClause().Types.Nodes {
			result = append(result, p.childNodeParser.CreateType(expression, context, nil))
		}
	}
	return result
}

// getProperties returns the parsed object properties. The second return value
// reports whether a required property resolved to never, which makes the
// whole type never (the upstream implementation returns undefined then).
func (p *InterfaceAndClassNodeParser) getProperties(node *ast.Node, context *Context) ([]*types.ObjectProperty, bool) {
	var members []*ast.Node
	for _, member := range node.Members() {
		if ast.IsConstructorDeclaration(member) {
			for _, param := range member.Parameters() {
				if ast.IsParameterPropertyDeclaration(param, param.Parent) {
					members = append(members, param)
				}
			}
		} else if ast.IsPropertySignatureDeclaration(member) || ast.IsPropertyDeclaration(member) {
			members = append(members, member)
		}
	}

	hasRequiredNever := false
	var properties []*types.ObjectProperty
	for _, member := range members {
		if !tsutils.IsPublic(member) || tsutils.IsStatic(member) || tsutils.IsNodeHidden(member) {
			continue
		}

		memberType := member.Type()

		// Use the type checker if the member has no explicit type.
		// Ignore members without an initializer; they have no useful type.
		if memberType == nil && member.Initializer() != nil {
			t := p.typeChecker.GetTypeAtLocation(member)
			memberType = p.typeChecker.TypeToTypeNode(t, node, nodeBuilderFlagsNoTruncation, synthesizedSymbols)

			// TypeToTypeNode returns a node that is detached from the AST, so
			// AnnotatedNodeParser cannot walk up to the member to read its
			// JSDoc. Point the synthesized node at the original declaration so
			// annotations such as the property description are preserved.
			// See upstream issue #1531.
			if memberType != nil {
				memberType.Parent = member
			}
		}

		if memberType == nil {
			continue
		}

		property := types.NewObjectProperty(
			p.getPropertyName(member.Name()),
			p.childNodeParser.CreateType(memberType, context, nil),
			member.QuestionToken() == nil,
		)

		if types.IsNeverLike(property.Type) {
			if property.Required {
				hasRequiredNever = true
			}
			continue
		}
		properties = append(properties, property)
	}

	if hasRequiredNever {
		return nil, true
	}
	return properties, false
}

func (p *InterfaceAndClassNodeParser) getAdditionalProperties(node *ast.Node, context *Context) any {
	for _, member := range node.Members() {
		if ast.IsIndexSignatureDeclaration(member) {
			if t := p.childNodeParser.CreateType(member.Type(), context, nil); t != nil {
				return t
			}
			return p.additionalProperties
		}
	}
	return p.additionalProperties
}

func (p *InterfaceAndClassNodeParser) getTypeId(node *ast.Node, context *Context) string {
	nodeType := "class"
	if ast.IsInterfaceDeclaration(node) {
		nodeType = "interface"
	}
	return nodeType + "-" + GetNodeKey(node, context)
}

func (p *InterfaceAndClassNodeParser) getPropertyName(propertyName *ast.Node) string {
	if propertyName.Kind == ast.KindComputedPropertyName {
		if symbol := tsutils.GetSymbolAtLocation(p.typeChecker, propertyName); symbol != nil {
			return symbol.Name
		}
	}
	return nodeText(propertyName)
}

// heritageClausesOf returns the heritage clauses of an interface or class
// declaration (empty otherwise).
func heritageClausesOf(node *ast.Node) []*ast.Node {
	var list *ast.NodeList
	switch node.Kind {
	case ast.KindInterfaceDeclaration:
		list = node.AsInterfaceDeclaration().HeritageClauses
	case ast.KindClassDeclaration:
		list = node.AsClassDeclaration().HeritageClauses
	}
	if list == nil {
		return nil
	}
	return list.Nodes
}
